Yosemite/Yosemite/Stores/AppSettings/InAppFeedbackCardVisibilityUseCase.swift

```diff
@@ -0,0 +1,115 @@
+/// - The user has not given feedback for more than 6 months ago.
+///
+struct InAppFeedbackCardVisibilityUseCase {
+    /// Errors returned by this UseCase.
+    enum InferenceError: Error {
+        case failedToInferInstallationDate
+        case unexpectedCalendarResult
+    }
+
+    private let fileManager: FileManager
+    private let calendar: Calendar
+
+    private let settings: GeneralAppSettings
+
+    init(settings: GeneralAppSettings, fileManager: FileManager = FileManager.default, calendar: Calendar = .current) {
+        self.settings = settings
+        self.fileManager = fileManager
+        self.calendar = calendar
+    }
+
+    /// Returns whether the In-app Feedback Card should be displayed.
+    ///
+    /// - Parameter currentDate The current date. This is only used for consistency in unit tests.
+    ///
+    func shouldBeVisible(currentDate: Date = Date()) throws -> Bool {
+        guard let installationDate = inferInstallationDate() else {
+            throw InferenceError.failedToInferInstallationDate
+        }
+
+        if try numberOfDays(from: installationDate, to: currentDate) < Constants.minimumInstallAgeInDays {
+            return false
+        }
+
+        guard let lastFeedbackDate = settings.lastFeedbackDate else {
+            return true
+        }
+
+        if try numberOfDays(from: lastFeedbackDate, to: currentDate) < Constants.feedbackFrequencyInDays {
+            return false
+        }
+
+        return true
+    }
+
+    /// Returns the total number of days between `from` and `to`.
+    private func numberOfDays(from: Date, to: Date) throws -> Int {
+        let components = [.day] as Set<Calendar.Component>
+        let dateComponents = calendar.dateComponents(components, from: from, to: to)
+        guard let days = dateComponents.day else {
+            throw InferenceError.unexpectedCalendarResult
+        }
+
+        return days
+    }
+
+    /// Retrieve the installation date.
+    ///
+    /// Checks both the date of `GeneralAppSettings.installationDate` and the creation date of the
+    /// Documents directory. The oldest of the two will be returned.
+    ///
+    /// We could simply just use the `GeneralAppSettings.installationDate` but we also have to
+    /// consider the users who have already installed before we started tracking that value.
+    ///
+    private func inferInstallationDate() -> Date? {
+        let documentDirCreationDate = creationDateOfDocumentDir()
+        let savedInstallationDate = settings.installationDate
+
+        if let documentDirCreationDate = documentDirCreationDate,
+            let savedInstallationDate = savedInstallationDate {
+            return min(documentDirCreationDate, savedInstallationDate)
+        } else if documentDirCreationDate != nil {
+            return documentDirCreationDate
+        } else if savedInstallationDate != nil {
+            return savedInstallationDate
+        } else {
+            return nil
+        }
```

Comment:
Do you think a switch could help a little bit with readability?

```swift
switch (creationDateOfDocumentDir(), settings.installationDate) {
    case let (creationDate?, installationDate?):
        return min(creationDate, installationDate)
    case let (creationDate?, nil):
        return creationDate
    case let (nil, instalationDate?):
        return instalationDate
    default:
        return nil
}
```

Reply:
Wow! I didn't know I could do that. Thanks for suggesting it! Changed in ed51de386ed1368aecd52a0bfa21a9d5d2a4f649.